opensearch-operator/pkg/reconcilers/tls.go

```diff
@@ -259,6 +259,54 @@ func (r *TLSReconciler) adminSecretName() string {
 	return r.instance.Name + "-admin-cert"
 }
 
+// CRITEO WORKAROUND
+func (r *TLSReconciler) shouldCreateGlobalTransportCert(ca tls.Cert) (bool, error) {
+	secret := &corev1.Secret{}
+	err := r.Get(r.ctx, types.NamespacedName{
+		Name:      r.instance.Name + "-transport-cert",
+		Namespace: r.instance.Namespace,
+	}, secret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.logger.Info("transport cert does not exist, creating")
+			r.logger.Info(r.instance.Name)
+			return true, nil
+		}
+		return false, err
+	}
+
+	data, ok := secret.Data[corev1.TLSCertKey]
+	if !ok {
+		return true, nil
+	}
+
+	deadline := 30 * 24 * time.Hour
+	if r.instance.Spec.Security.Tls.Transport.RegenDeadlineDays > 0 {
+		deadline = time.Duration(r.instance.Spec.Security.Tls.Transport.RegenDeadlineDays*24) * time.Hour
+	}
+
+	validator, err := tls.NewCertValidater(data, tls.WithExpiryThreshold(deadline))
+	if err != nil {
+		return false, err
+	}
+
+	if validator.IsExpiringSoon() {
+		r.logger.Info("transport cert is expiring soon, recreating")
+		return true, nil
+	}
+
+	verified, err := validator.IsSignedByCA(ca)
+	if err != nil {
+		return false, err
+	}
+
+	if !verified {
+		r.logger.Info("transport cert is not signed by CA, recreating")
+	}
+
+	return !verified, nil
+}
+
 func (r *TLSReconciler) handleTransportGenerateGlobal() error {
```

Comment:
please add unit test to cover your change

Reply:
I see a few impediments to this (tls tests are in `opensearch-operator/pkg/reconcilers/tls_test.go`):

- opensearch operator tests rely on kubernetes `recorder` package to mock API calls. This makes kube API unit testing fairly straightforward, however cert parsing (extracting expiry date and checking against CA) does not use kube API
- for certificate unit testing, there is a mock API `opensearch-k8s-operator/opensearch-operator/pkg/helpers/test-helpers.go` but the function are all stubs:
    - cert/key data is hardcoded to the filename (eg `[]byte("ca.crt")`) and are not valid certs
    - cert validator class and cert generator functions are obviously not mocked, either

Moreover, it seems that upstream has opted against unit-testing admin cert auto-update (they do check if certs are created when they don't exist, though), which is almost identical to what my change does to transport cert.